### PR TITLE
storage: persist proposed lease transfers to disk

### LINKED ISSUE #7996: storage: lease transfer must be persisted to disk
Or a node which transfers its lease away might restart rapidly and use a tail end of its lease while its lease was really shortened.

The naive fix (which might well be the right one) is to persist the expiration date of the transferred lease to disk (keyed by RangeID) and, when starting the process, sitting ducks until `the local HLC clock - max_offset` reports a higher timestamp.

There is also the somewhat related issue that the hlc clock itself may jump backwards. For that reason alone, we should already sleep `max_offset` when starting the process.

### PR DESCRIPTION
**Pull request by andreimatei**
... so we don't use those leases after a restart

fixes #7996

cc @cockroachdb/stability @petermattis

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9523)

### CODE REVIEW

File: keys/constants.go [hunk 1]
```diff
@@ -80,6 +80,9 @@ var (
 	// localStoreGossipSuffix stores gossip bootstrap metadata for this
 	// store, updated any time new gossip hosts are encountered.
 	localStoreGossipSuffix = []byte("goss")
+	// localStoreSafeStartSuffix stores the minimum timestamp when it's safe for
+	// the store to start serving after a restart.
+	localStoreSafeStartSuffix = []byte("safe-start")
```
Comment (bdarnell) [inline keys/constants.go]: All suffixes are supposed to be 4 characters (see `localSuffixLength` above).

File: storage/store.go [hunk 2]
```diff
@@ -448,6 +448,9 @@ type Store struct {
 		// raft.
 		droppedPlaceholders int32
 	}
+
+	// safeStartMu serializes access to the engine's "safe start" key.
+	safeStartMu syncutil.Mutex
```
Comment (bdarnell) [inline storage/store.go]: Why does this get a new mutex?

File: server/server.go [hunk 3]
```diff
@@ -492,6 +492,31 @@ func (s *Server) Start(ctx context.Context) error {
...
+		return nil
+	})
+	if err != nil {
+		return errors.Wrap(err, "error reading stores' safe start")
+	}
+	if s.clock.Now().Less(safeStart) {
+		log.Infof(ctx, "waiting for engines' safe start timestamp...")
+		time.Sleep(
+			// Add 1 nanosecond because we're ignoring the logical part.
+			time.Duration(safeStart.WallTime-s.clock.PhysicalNow()+1) * time.Nanosecond)
```
Comment (bdarnell) [inline server/server.go]: Don't call `clock.{Physical,}Now()` more than once here.

**Review by bdarnell — COMMENTED** [review]
Lease transfers are a per-range operation; blocking the entire store on restart seems unnecessarily coarse (especially if we start transferring all our leases away on shutdown; then the worst case of having to wait for almost the full lease duration (9s) will be common). I'd rather store this at the replica level instead of the store level. 

We could almost do it by modifying our persisted version of the lease, but that is defined as replicated state, so we need to introduce a new unreplicated range-id-local key.

**Comment by andreimatei** [comment]
@bdarnell yeah, this mechanism is course (and this is explained in the comments in this commit), but the alternative seemed complex. Not only would we need to store per-range (or per-lease) local keys, but we'd also need a way to not use the respective individual leases after a restart (and this mechanism should probably be different than what's used by the server for not using the transferred lease before a restart - the `replica.pendingLeaseRequest` struct). The idea of just marking the local lease as tainted was considered originally when I introduced the lease transfers, but, as you've said, the idea of applying a non-replicated change to this otherwise replicated state seemed like a bad idea.

I was punting on doing anything better for the time when/if non-expiring leases become a reality. In the meantime, note that we only wait out the lease duration when a server restarts immediately (which should not be common, I think). Also a server takes a while to shutdown, so depending on how we do the lease transfers on shutdown, that might also take care of some of the wait duration.

So, I dunno... If you don't feel strongly, I'd go ahead with this change to get these transfers be correct and improve... later.

**Comment by bdarnell** [comment]
Storing this per-replica doesn't seem significantly more complicated than doing it per-store. You just need a new unreplicated member of ReplicaState, which can be loaded at startup just like the previous lease is. You'd check this lease transfer timestamp at the same time you check whether the current lease covers the current timestamp. I'd much rather keep this at the replica level than introduce something entirely new at the store level that interacts with leases.

**Comment by tschottdorf** [comment]
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523)*

**Comment by tamird** [comment]
Reviewed 6 of 8 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[server/server.go, line 510 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSbn4WeZvuoz4hhaIij:-KSbn4WfJiC0dXkWHx-l:b-g2wx8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/server/server.go#L510)):*
> ```Go
> 	})
> 	if err != nil {
> 		return errors.Wrap(err, "error reading stores' safe start")
> ```

this wrapping should happen inside the closure so we know which store.

---

*[server/server.go, line 513 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSbnDq7GNZtYoKLISGy:-KSbnDq7GNZtYoKLISGz:b-d6duip) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/server/server.go#L513)):*
> ```Go
> 	}
> 	if s.clock.Now().Less(safeStart) {
> 		log.Infof(ctx, "waiting for engines' safe start timestamp...")
> ```

is it the engines' or the stores'?

---

*[server/server.go, line 517 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSbnMe3i0ZTNi_STiXc:-KSbnMe3i0ZTNi_STiXd:b9evwfy) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/server/server.go#L517)):*
> ```Go
> 			// Add 1 nanosecond because we're ignoring the logical part.
> 			time.Duration(safeStart.WallTime-s.clock.PhysicalNow()+1) * time.Nanosecond)
> 		log.Infof(ctx, "waiting for engines' safe start timestamp... done")
> ```

not sure you need the common prefix here - the server isn't running yet, so there's nothing else in the logs.

---

*[storage/replica_range_lease.go, line 292 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSbmsOAQ4ofd--3og2R:-KSbmsOAQ4ofd--3og2S:b59qcha) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/storage/replica_range_lease.go#L292)):*
> ```Go
> 
> 		if err := r.store.MaybeUpdateSafeStart(ctx, lease); err != nil {
> 			return nil, nil, errors.Wrap(err, "error persisting lease transfer")
> ```

"error updating safe start"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523)*

**Comment by andreimatei** [comment]
ok, will rewrite

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-:-KScbHLh7JofZeoIxjaC:b9y8wj)*

**Comment by tschottdorf** [comment]
Assigned `stability` label to mark for supervised merge.

**Comment by andreimatei** [comment]
I've rewritten the thing. Now, each replica's lease transfer state is persisted to disk in such a way that, upon restart, it can be almost perfectly reconstructed. So there's no longer a difference between how we're checking if we're in the middle of a transfer during regular running versus immediately after a restart.
This required an interaction between the raft machinery and the `pendingLeaseRequest` struct - whereas before one would request a new lease (new, extension or transfer) and then wait on the `replica.Send` channel to see if the new lease has applied or not (an ephemeral mechanism that cannot survive a restart), now it will wait on a channel that's signaled by the commit trigger that applies the new lease. This arguably broke an encapsulation of the logic in the `pendingLeaseRequest` struct, but, besides serving the purpose of the PR, it's also more correct (before, it was unclear what would happen if, say, someone's waiting on a lease proposal, but another lease is applied in the meantime). The change is also in the spirit of the `Replica`, having reconstructible state, which `pendingLeaseRequest` was going against.

@spencerkimball, since you insisted on the idea of converging in-memory and on-disk state, I think you get the review. But say if you're too busy.

`r7` is the main commit of interest, the rest are misc. I'll probably need to add more tests that mimic a restart, but wanna get some validation.

CC @raduberinde for `r6` - expending the `TestCluster` shim.

---

Review status: 0 of 27 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[keys/constants.go, line 85 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSSHW4NBXZxZkJTOQ57-r1-85:-KTpOvweVAAPaIzmJZ-m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/keys/constants.go#L85)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> All suffixes are supposed to be 4 characters (see `localSuffixLength` above). 
</details>
Done.

---

*[storage/replica_range_lease.go, line 292 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSbmsOAQ4ofd--3og2R:-KTpP14RdwhJNPs-SxnB:b-ynslp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/storage/replica_range_lease.go#L292)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "error updating safe start"
</details>
it's gone

---

*[storage/store.go, line 453 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSSHW4OaY9llFKpc9Lj-r1-453:-KTpP31ybTzS2t2z55wy:bb7gzl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/storage/store.go#L453)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why does this get a new mutex?
</details>
gone

---

*[server/server.go, line 510 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSbn4WeZvuoz4hhaIij:-KTpP4mJceO1d0jAE7gk:bb7gzl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/server/server.go#L510)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this wrapping should happen inside the closure so we know which store.
</details>
gone

---

*[server/server.go, line 513 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSbnDq7GNZtYoKLISGy:-KTpP5RGFfkvpDcLajOz:bb7gzl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/server/server.go#L513)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is it the engines' or the stores'?
</details>
gone

---

*[server/server.go, line 516 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSSHW4NBXZxZkJTOQ59-r1-516:-KTpP73ichO--Mon0rIw:bb7gzl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/server/server.go#L516)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Don't call `clock.{Physical,}Now()` more than once here. 
</details>
gone

---

*[server/server.go, line 517 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KSbnMe3i0ZTNi_STiXc:-KTpP7ezkg83hGO4cVtY:bb7gzl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/91d83b3a687ce77b0020e7da76015ede3b9ebbbf/server/server.go#L517)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not sure you need the common prefix here - the server isn't running yet, so there's nothing else in the logs.
</details>
gone

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-:-KTpP8xfc-h3PMMetqxl:br09v93)*

**Comment by bdarnell** [comment]
Reviewed 2 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 11 of 12 files at r6, 12 of 12 files at r7.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[keys/constants.go, line 118 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrF4Kp-FtCqBz7A5Lc:-KTrF4Kp-FtCqBz7A5Ld:bteb9e0) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/keys/constants.go#L118)):*
> ```Go
> 	LocalRangeLeaseSuffix = []byte("rll-")
> 	// LocalNextLeaseSuffix is the suffix for a range's "next lease".
> 	LocalNextLeaseSuffix = []byte("rnl-")
> ```

Move this after "rlla" below. They're alphabetized by value to make it easier to spot conflicts. 

---

*[storage/client_test.go, line 317 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrN5ku7RUfyO3UYO3c:-KTrN5kvt-MV5mT4zWiR:b-hkd8mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/client_test.go#L317)):*
> ```Go
> 		wg.Add(len(m.stoppers))
> 		for i, s := range m.stoppers {
> 			go func(i int, s *stop.Stopper) {
> ```

This new argument is unused.

---

*[storage/client_test.go, line 357 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrNA2kPXmpzEO2UjF0:-KTrNA2kPXmpzEO2UjF1:b-qfekik) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/client_test.go#L357)):*
> ```Go
> 			m.t.Error("timed out during shutdown")
> 		} else {
> 			panic(fmt.Sprintf("timed out during shutdown. current time: %s.", timeutil.Now()))
> ```

This could be a log.Fatal instead and then there'd be a timestamp automatically. 

---

*[storage/log_test.go, line 151 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrNW0rRG_WGg4D2Ar4:-KTrNW0rRG_WGg4D2Ar5:b-n3nnfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/log_test.go#L151)):*
> ```Go
> 	// Stores (or a creative usage of VisitStores could suffice).
> 	testStore, err := s.(*server.TestServer).Stores().GetStore(roachpb.StoreID(1))
> 	store := testStore.(*storage.Store)
> ```

Its unfortunate that all these explicit casts are necessary here in the storage package where all the types are accessible. 

---

*[storage/replica_command.go, line 1747 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrRqkhQxuS9o34ers5:-KTrRqkhQxuS9o34ers6:bntyqda) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/replica_command.go#L1747)):*
> ```Go
> 	// replays.
> 	// TODO(andrei): I've seen replays of a LeaseRequest in a test, happening very
> 	// predictably. Is this expected?
> ```

Yes, I think that's expected. But why do we need to handle this specially, instead of treating it as an "extension" that doesn't actually move the expiration forward?

---

*[storage/replica_range_lease.go, line 114 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrTLfeT7Oj6bEm89Lo:-KTrTLfeT7Oj6bEm89Lp:b-jhsc92) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/replica_range_lease.go#L114)):*
> ```Go
> 		// request this signal corresponds too. It might not correspond to the
> 		// request the current waiters are waiting on, since a signal delivering a
> 		// success can race with a signal signaling an error for the same request.
> ```

Why would the same request both succeed and fail? There are lease index and reproposal failures, but those don't make it far enough to call SignalLeaseApplied. 

---

*[storage/replica_range_lease.go, line 131 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrUVLo44o4txabGFsv:-KTrUVLo44o4txabGFsw:b-2ndj3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/replica_range_lease.go#L131)):*
> ```Go
> 		// lease to be used for serving.
> 		if isExtension := lease.Replica.ReplicaID == p.replica.mu.replicaID; isExtension {
> 			p.WipeNextLease(ctx)
> ```

I don't see any tests that show that this makes a difference. What kind of failures could reach this point while the old lease is still active, and what would be the consequences of handling extensions in the same way as transfers (i.e. conservatively allowing NextLease to remain)?

---

*[storage/replica_range_lease.go, line 145 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrVIRY245OfZAFmDAE:-KTrVIRY245OfZAFmDAF:b-wsy7je) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/replica_range_lease.go#L145)):*
> ```Go
> 		ctx, p.replica.store.engine, p.replica.RangeID, roachpb.Lease{})
> 	if err != nil {
> 		// TODO(andrei): What should we do for such replica corruption errors?
> ```

We have the Replica.maybeSetCorrupt method (soon to move to Store). 

---

*[storage/replica_range_lease.go, line 154 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrVVMZ59PBdz7H1Wxj:-KTrVVMZ59PBdz7H1Wxk:box2qqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/replica_range_lease.go#L154)):*
> ```Go
> // applies a lease (the lease the caller has just proposed, or a completely
> // different one).
> func (p *pendingLeaseRequest) getWaitChan(ctx context.Context) chan leaseOrErr {
> ```

If this can easily return a `<-chan`, it should.

---

*[storage/replica_range_lease.go, line 162 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrVgTorwyVOLOYdGZJ:-KTrVgTorwyVOLOYdGZK:berwikz) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/replica_range_lease.go#L162)):*
> ```Go
> 
> // notifyWaiters sends a notification to the current batch of waiters. It then
> // clears the list of waiters. Callers should also call WipeNextLease atomically
> ```

What does "atomically" mean here? Under some lock? Which one?

---

*[storage/replica_range_lease.go, line 252 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrWnF6cit14XruGvyY:-KTrWnF6cit14XruGvyZ:bzgyrxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/replica_range_lease.go#L252)):*
> ```Go
> 			// NOTE(andrei): We could return an error to the caller here since we
> 			// haven't sent the request yet, but we can't similarly handle the same
> 			// kind of corruption error in GetWaitChan(), so we just panic here too
> ```

getWaitChan doesn't have a panic.

---

*[storage/replica_range_lease.go, line 309 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-KTrX4W74WHx2-cokoeh:-KTrX4W74WHx2-cokoei:b4nixd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/81eee1d6684430e72e619ca57b97d4a5a232fe14/storage/replica_range_lease.go#L309)):*
> ```Go
> // LeaderLease.
> //
> // Note that the pendingLeaseRequest, and hence this method, is unaware of any
> ```

Is this still true?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523)*

**Comment by spencerkimball** [comment]
@andreimatei do you still want me to review this?

**Comment by andreimatei** [comment]
Ben seems to be on it. Probably enough.

**Comment by andreimatei** [comment]
So something came up when discussing with @tschottdorf , on the sides of his comments in #6290 about using leases after restarts in a proposer-evaluated KV world.
It seems that, even today, it's incorrect to serve reads after a restart given that the command queue has been wiped on restart - the serialization of reads with writes that might be in flight is gone and that's a correctness problem. So it looks like, after a restart, we can't use any of the leases we had before the restart (as opposed to not using the ones that were in the process of being transferred, as this PR is trying to do).
So I'm wondering if we should give up this PR, or at least give up on relying on it for protecting against anything on restart, and revert to a blunt way of sitting out *all* old leases when upon restart.

@bdarnell @spencerkimball any opinions?

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9523#-:-KUNonrQMBS_2oELj5AX:b-t1uwh5)*

**Comment by spencerkimball** [comment]
I would in general be in favor of the blunt approach. Complexity scares me. And with the epoch-based range leases, it becomes less and less likely that any ranges on a restarted node will be reusing existing leases anyway.

But in general, if a node is restarted, none of its ranges should ever be allowed to use pre-existing leases until all committed log entries have been applied to the finite state machine. So the command queue issue which you're worried about should be a moot point.

**Comment by bdarnell** [comment]
Yeah, I think I'd abandon this change and instead make sure that we only use leases that were requested by the current process. 

> But in general, if a node is restarted, none of its ranges should ever be allowed to use pre-existing leases until all committed log entries have been applied to the finite state machine. So the command queue issue which you're worried about should be a moot point.

I'm not sure what you're trying to say here; we certainly don't enforce this today.

**Comment by spencerkimball** [comment]
@bdarnell yes we don't enforce this today. In more detail: if you restart with the lease still unexpired, and serve a read immediately for a key which has a pending, committed change in the raft log which hasn't yet been applied, you'll have a problem. I'm not entirely sure the scenario is possible, but I think it could be unless we apply all committed commands at raft group load time.

The simpler approach we're now advocating of outlawing use of leases not from the current process will fix this potential consistency vulnerability.

**Comment by andreimatei** [comment]
I've opened a new PR (#10211) instead of this one, addressing the original motivation (don't reuse transferred leases) and more.
I'll now see what of this PR deserves salvaging.

**Comment by tschottdorf** [comment]
Unassigning due to not being able to review this past 11/6.

**Comment by andreimatei** [comment]
closing in favor of #10420

**Pull request closed without merging**